homeassistant/components/onvif/camera.py

```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
+    async def async_perform_ptz(
+        self, pan, tilt, zoom, distance, speed, move_mode, continuous_duration
+    ):
         """Perform a PTZ action on the camera."""
         if self._ptz_service is None:
             _LOGGER.warning("PTZ actions are not supported on camera '%s'", self._name)
             return
 
         if self._ptz_service:
-            pan_val = 1 if pan == DIR_RIGHT else -1 if pan == DIR_LEFT else 0
-            tilt_val = 1 if tilt == DIR_UP else -1 if tilt == DIR_DOWN else 0
-            zoom_val = 1 if zoom == ZOOM_IN else -1 if zoom == ZOOM_OUT else 0
-            req = {
-                "Velocity": {
-                    "PanTilt": {"_x": pan_val, "_y": tilt_val},
-                    "Zoom": {"_x": zoom_val},
-                }
-            }
+            pan_val = (
+                distance if pan == DIR_RIGHT else -distance if pan == DIR_LEFT else 0
+            )
+            tilt_val = (
+                distance if tilt == DIR_UP else -distance if tilt == DIR_DOWN else 0
+            )
+            zoom_val = (
+                distance if zoom == ZOOM_IN else -distance if zoom == ZOOM_OUT else 0
+            )
+            speed_val = speed
+            _LOGGER.debug(
+                "Calling %s PTZ | Pan = %d | Tilt = %d | Zoom = %d | Speed = %d",
+                move_mode,
+                pan_val,
+                tilt_val,
+                zoom_val,
+                speed_val,
+            )
             try:
-                _LOGGER.debug(
-                    "Calling PTZ | Pan = %d | Tilt = %d | Zoom = %d",
-                    pan_val,
-                    tilt_val,
-                    zoom_val,
-                )
-
-                await self._ptz_service.ContinuousMove(req)
+                req = self._ptz_service.create_type(move_mode)
+                req.ProfileToken = await self.async_obtain_profile_token()
+                if move_mode == "ContinuousMove":
+                    req.Velocity = {
+                        "PanTilt": {"x": pan_val, "y": tilt_val},
+                        "Zoom": {"x": zoom_val},
+                    }
+
+                    await self._ptz_service.ContinuousMove(req)
+                    sleep(continuous_duration)
+                    req = self._ptz_service.create_type("Stop")
+                    req.ProfileToken = await self.async_obtain_profile_token()
+                    await self._ptz_service.Stop({"ProfileToken": req.ProfileToken})
+                elif move_mode == "RelativeMove":
+                    req.Translation = {
+                        "PanTilt": {"x": pan_val, "y": tilt_val},
+                        "Zoom": {"x": zoom_val},
+                    }
+                    req.Speed = {
+                        "PanTilt": {"x": speed_val, "y": speed_val},
+                        "Zoom": {"x": speed_val},
+                    }
+                    await self._ptz_service.RelativeMove(req)
+                elif move_mode == "AbsoluteMove":
```

Comment:
I recommend using the constant you defined instead of the raw string.

Reply:
Thanks @djpremier

What I was pointing out in my previous post is that the description/documentation additions for **move_mode** in camera/services.yaml does not match **move_mode** in onvif/camera.py code.  EG: services.yaml "AbsoluteMode" verses camera.py "AbsolueMove"

Even so, setting move_mode to AbsoluteMove or RelativeMove **does not work** for my FOSCAM ONVIF camera.  The resulting logs were included in my previous post.